Cloud functions and rules

rhelmer opened the pull request:
This depends on https://github.com/mozilla-rally/rally-web-platform/pull/100, addresses #55, closes #84, and closes #69.

This implements everything from the cloud function proposal required to create new users:

- [x] `addRallyUserToFirestore` cloud  function  - Auth `onCreate` to create all firestore docs
- [x] new user studies subcollection, removing old enrolledStudies map in user doc
- [x] move svelte app from enrolledStudies to user studies subcollection
- [x] update rally-sdk in test extensions to use user studies subcollection
- [x] add cloud firestore rules to enforce the above 

This also adds an `npm run watch` command which runs the Firebase emulators and watches for changes to rules, cloud functions, static data (studies), and the website, which makes it much easier to make+test changes like this.

rhelmer commented:
OK so the test is a lot more reliable overall now, still hitting odd situations where this notification is obscuring the Leave button, somehow:

![image](https://user-images.githubusercontent.com/61412/131589409-5d3fc05d-ffbd-435a-a3c7-0eb2d2d093af.png)

rhelmer commented:
This seems to work well with the latest rally-sdk (https://github.com/mozilla-rally/rally-sdk/commit/d16a00014e4fe84fc0e2a8feb66043da59e0c5e8 as of this writing), the tests seem more reliable so far.

We still need to implement Glean pings and deletion requests, this one is already getting quite large so I'd like to do that as a separate bit of work.

rhelmer commented:
> We still need to implement Glean pings and deletion requests, this one is already getting quite large so I'd like to do that as a separate bit of work.

Filed https://github.com/mozilla-rally/rally-web-platform/issues/108 to track that.

hamilton commented on package.json:
```diff
@@ -4,9 +4,13 @@
-    "build:test:integration": "npm run pre-build && svelte-kit build -- --config-integration-test-mode",
+    "build:web:emulator": "npm run pre-build && svelte-kit build -- --config-integration-test-mode",
+    "watch:functions": "watch 'npm run build:functions && sleep 1 && npm run load:data' ./functions/src",
+    "watch:web": "watch 'svelte-kit dev -- --config-integration-test-mode' ./src",
```
Should we deprecate `dev` here / everywhere in favor of `watch`, or vice versa?

hamilton reviewed (approved):
This is great – so many improvements! Had a few comments but looks good to me

Carla-Moz reviewed (approved):
Working for me and the README looks greats. Thanks!

rhelmer commented:
OK I've addressed review comments, thanks! I buttoned up the last few things we discussed in Slack, and  I removed the single-study hardcoding and it should in theory support multiple studies (we have tests with a single study, I haven't tried multiple yet).

This wfm with `npm run dev` under the VSCode debugger, and tests are passing on Circle now. The only thing I am going to defer is to not generate tokens until the SDK specifically asks (this requires an SDK change, https://github.com/mozilla-rally/rally-sdk/issues/13)
